zrml/neo-swaps/src/lib.rs

```diff
@@ -0,0 +1,903 @@
+                // `amount_out_minus_fees` units of collateral to `who`. The fees automatically end
+                // up in the pool.
+                let amount_out = pool.calculate_swap_amount_out_for_sell(asset_in, amount_in)?;
+                // Beware! This transfer happen _after_ calculating `amount_out`:
+                T::AssetManager::transfer(asset_in, &who, &pool.account_id, amount_in)?;
+                T::CompleteSetOperations::sell_complete_set(
+                    pool.account_id.clone(),
+                    market_id,
+                    amount_out,
+                )?;
+                let FeeDistribution {
+                    remaining: amount_out_minus_fees,
+                    swap_fees: swap_fee_amount,
+                    external_fees: external_fee_amount,
+                } = Self::distribute_fees(market_id, pool, amount_out)?;
+                ensure!(amount_out_minus_fees >= min_amount_out, Error::<T>::AmountOutBelowMin);
+                T::AssetManager::transfer(
+                    pool.collateral,
+                    &pool.account_id,
+                    &who,
+                    amount_out_minus_fees,
+                )?;
+                for (&asset, balance) in pool.reserves.iter_mut() {
+                    if asset == asset_in {
+                        *balance = balance.checked_add(&amount_in).ok_or(Error::<T>::MathError)?;
+                    }
+                    *balance = balance.checked_sub(&amount_out).ok_or(Error::<T>::MathError)?;
+                }
+                let new_price = pool.calculate_spot_price(asset_in)?;
+                ensure!(
+                    new_price >= MIN_SPOT_PRICE.saturated_into(),
+                    Error::<T>::SpotPriceBelowMin
+                );
+                Self::deposit_event(Event::<T>::SellExecuted {
+                    who: who.clone(),
+                    market_id,
+                    asset_in,
+                    amount_in,
+                    amount_out,
+                    swap_fee_amount,
+                    external_fee_amount,
+                });
+                Ok(())
+            })
+        }
+
+        #[require_transactional]
+        fn do_join(
+            who: T::AccountId,
+            market_id: MarketIdOf<T>,
+            pool_shares_amount: BalanceOf<T>,
+            max_amounts_in: Vec<BalanceOf<T>>,
+        ) -> DispatchResult {
+            ensure!(pool_shares_amount != Zero::zero(), Error::<T>::ZeroAmount);
+            let market = T::MarketCommons::market(&market_id)?;
+            ensure!(market.status == MarketStatus::Active, Error::<T>::MarketNotActive);
+            Self::try_mutate_pool(&market_id, |pool| {
+                // FIXME Round up to avoid exploits.
+                let ratio = bdiv(
+                    pool_shares_amount.saturated_into(),
+                    pool.liquidity_shares_manager.total_shares()?.saturated_into(),
+                )?;
```

Comment:
Logic seems correct, based on the pdf doc, users puts in `λrk` outcome tokens and receive `λq` pool shares, thus `λ = sum(qk) / sum(rk)` as the more pool shares are received the higher the amount of tokens to put into the pool.

Reply:
What does `qk` refer to here?

Reply:
`qk` is the amount of pool shares received for adding liquidity of the outcome token `rk`. `pool_shares_amount = sum(qk)`. Probably a strange way to view it, as adding liquidity happens atomically for all outcome tokens at once.